compiler/src/iree/compiler/ExternalInterfaces/EncodingExternalModels.cpp

```diff
@@ -0,0 +1,128 @@
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Builders.h"
+
+namespace mlir::iree_compiler {
+namespace {
+
+struct ContractionAttrPropagationInterface
+    : public IREE::Encoding::EncodingPropagationAttrInterface::ExternalModel<
+          ContractionAttrPropagationInterface, IREE::Encoding::MatmulKAttr> {
+  bool isPropagable(Attribute attr, Value target) const {
+    auto encoding = dyn_cast<IREE::Encoding::MatmulKAttr>(attr);
+    if (!encoding) {
+      return false;
+    }
+    if (auto collapseOp = target.getDefiningOp<tensor::CollapseShapeOp>()) {
+      ArrayRef<int32_t> kDims = encoding.getKDims().asArrayRef();
+
+      // Bail out if it is not propagable.
+      // TODO: Relax the check to allow transforming innermost reduction
+      // dimensions. We need to revisit the matmul_k encoding semantic.
+      SmallVector<ReassociationIndices, 4> reassociationMaps =
+          collapseOp.getReassociationIndices();
+      for (int32_t k : kDims) {
+        if (reassociationMaps[k].size() != 1) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return false;
+  }
+
+  FailureOr<IREE::Encoding::PropagationEncoding>
+  generateEncodings(Attribute attr, Value target) const {
+    auto encoding = dyn_cast<IREE::Encoding::MatmulKAttr>(attr);
+    if (!encoding) {
+      return failure();
+    }
+    if (auto collapseOp = target.getDefiningOp<tensor::CollapseShapeOp>()) {
+      ArrayRef<int32_t> kDims = encoding.getKDims().asArrayRef();
+      SmallVector<ReassociationIndices, 4> reassociationMaps =
+          collapseOp.getReassociationIndices();
+      // Get a mapping from original iteration space to expanded iteration
+      // space.
+      SmallVector<int32_t> newKDims;
+      for (int32_t kDim : kDims) {
+        newKDims.append(reassociationMaps[kDim].begin(),
+                        reassociationMaps[kDim].end());
+      }
+      MLIRContext *ctx = collapseOp.getContext();
+      auto resultEncodingAttr = IREE::Encoding::MatmulKAttr::get(ctx, newKDims);
+
+      IREE::Encoding::PropagationEncoding propEncoding;
+      propEncoding.operandEncodings.push_back(resultEncodingAttr);
+      // The result encoding will be the same as the encoding
+      // present in the set encoding operation.
+      propEncoding.resultEncodings.push_back(encoding);
+      return propEncoding;
+    }
+    return failure();
+  }
+};
+
+struct ContractionOpPropagationInterface
+    : public IREE::Encoding::EncodingPropagationOpInterface::ExternalModel<
+          ContractionOpPropagationInterface, tensor::CollapseShapeOp> {
+  FailureOr<IREE::Encoding::PropagationResult>
+  propagateEncoding(Operation *op, RewriterBase &builder,
+                    IREE::Encoding::PropagationEncoding encodings,
+                    OpResult opResult) const {
+
+    auto operandEncodings = encodings.operandEncodings;
+    auto resultEncodings = encodings.resultEncodings;
+
```

Comment:
style nit: remove the blank lines. Please also remove the blank lines in the below if body.

nit: You can declare `Location loc = op->getLoc()` which makes the below code simpler. We will use the variable for other future code, so it makes sense to me.